fcrepo-clients/src/main/java/edu/unc/lib/dl/acl/fcrepo4/InheritedAclFactory.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2017 The University of North Carolina at Chapel Hill
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.unc.lib.dl.acl.fcrepo4;
+
+import static edu.unc.lib.dl.acl.util.PrincipalClassifier.getPatronPrincipals;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import edu.unc.lib.dl.acl.service.PatronAccess;
+import edu.unc.lib.dl.acl.util.UserRole;
+import edu.unc.lib.dl.fedora.ContentPathFactory;
+import edu.unc.lib.dl.fedora.PID;
+
+/**
+ * Factory which provides access control details that apply to particular
+ * objects, either via direct definition or inherited from parent objects
+ * 
+ * @author bbpennel
+ *
+ */
+public class InheritedAclFactory implements AclFactory {
+
+	private ObjectAclFactory objectAclFactory;
+
+	private ContentPathFactory pathFactory;
+
+	private ObjectPermissionEvaluator objectPermissionEvaluator;
+
+	@Override
+	public Map<String, Set<String>> getPrincipalRoles(PID target) {
+
+		// Retrieve the path of objects up to and including the target
+		List<PID> path = getPidPath(target);
+
+		Map<String, Set<String>> inheritedPrincRoles = new HashMap<>();
+		Set<String> patronPrincipals = null;
+
+		// Iterate through each step in the path except for the root content node
+		int depth = 0;
+		for (; depth < path.size(); depth++) {
+			PID pathPid = path.get(depth);
+
+			// For the first two objects (unit, collection), staff roles should be considered
+			if (depth < 2) {
+
+				Map<String, Set<String>> objectPrincipalRoles = objectAclFactory.getPrincipalRoles(pathPid);
+
+				// Add this object's principals/roles to the result
+				mergePrincipalRoles(inheritedPrincRoles, objectPrincipalRoles);
```

Comment:
This is a random spot for this comment, but I'm thinking it could be helpful to have a basic definition somewhere in the code of what principals are, since the terms is being used with a specific, technical meaning that may not be immediately clear to future devs.